src/werkzeug/debug/shared/debugger.js

```diff
@@ -1,210 +1,319 @@
+            btn.disabled = true;
+
+            fetch(`${document.location.pathname}?__debugger__=yes&cmd=pinauth&pin=${pin}&s=${encodedSecret}`)
+                .then(res => res.json())
+                .then(({ auth, exhausted }) => {
+                    if (auth) {
+                        EVALEX_TRUSTED = true;
+                        fadeOut(document.getElementsByClassName("pin-prompt")[0]);
+                    } else {
+                        alert(`Error: ${exhausted
+                        ? "too many attempts.  Restart server to retry."
+                        : "incorrect pin"}`)
+                    }
+                })
+                .catch(err => {
+                    alert('Error: Could not verify PIN.  Network error?');
+                    console.error(err);
+                })
+                .finally(() => btn.disabled = false);
+        }, false);
 }
 
 function promptForPin() {
-  if (!EVALEX_TRUSTED) {
-    $.ajax({
-      url: document.location.pathname,
-      data: {__debugger__: 'yes', cmd: 'printpin', s: SECRET}
+    if (!EVALEX_TRUSTED) {
+        const encodedSecret = encodeURIComponent(SECRET);
+        fetch(`${document.location.pathname}?__debugger__=yes&cmd=printpin&s=${encodedSecret}`);
+        const pinPrompt = document.getElementsByClassName("pin-prompt")[0];
+        fadeIn(pinPrompt);
+        document.querySelector('.pin-prompt input[name="pin"]').focus();
+    }
+}
+
+/**
+ * Helper function for shell initialization
+ */
+function openShell(consoleNode, target, frameID) {
+    promptForPin();
+    if (consoleNode) {
+        slideToggle(consoleNode);
+        return consoleNode;
+    }
+
+    let historyPos = 0;
+    let history = [''];
+
+    const consoleElement = createConsole();
+    const output = createConsoleOutput();
+    const form = createConsoleInputForm();
+    const command = createConsoleInput();
+
+    target.parentNode.appendChild(consoleElement);
+    consoleElement.append(output);
+    consoleElement.append(form);
+    form.append(command);
+    command.focus();
+    slideToggle(consoleElement);
+
+    form.addEventListener('submit', e => {
+        handleConsoleSubmit(e, command, frameID).then(consoleOutput => {
+            output.append(consoleOutput);
+            command.focus();
+            consoleElement.scrollTo(0, consoleElement.scrollHeight);
+            let old = history.pop();
+            history.push(command.value);
+            if (typeof old != 'undefined') {
+                history.push(old);
+            }
+            historyPos = history.length - 1;
+            command.value = '';
+        });
     });
-    $('.pin-prompt').fadeIn(function() {
-      $('.pin-prompt input[name="pin"]').focus();
+
+    command.addEventListener('keydown', function(e) {
+        if (e.key === 'l' && e.ctrlKey) {
+            output.innerText = '--- screen cleared ---';
+        } else if (e.charCode === 0 && (e.keyCode === 38 || e.keyCode === 40)) {
```

Comment:
Nowadays there are more modern ways to check for this, instead of obscure numeric key codes: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key